sdk/python/feast/infra/compute_engines/ray/feature_builder.py

```diff
@@ -0,0 +1,314 @@
+
+        transformation = feature_transformation or udf
+        if not transformation:
+            raise ValueError(f"No feature transformation found for {view.name}")
+
+        node = RayTransformationNode(
+            name="transformation",
+            transformation=transformation,
+            config=self.config,
+        )
+        for input_node in input_nodes:
+            node.add_input(input_node)
+
+        self.nodes.append(node)
+        return node
+
+    def build_output_nodes(self, view, final_node):
+        """Build the output node for writing processed features."""
+        node = RayWriteNode(
+            name="output",
+            feature_view=view,
+            inputs=[final_node],
+        )
+
+        self.nodes.append(node)
+        return node
+
+    def build_validation_node(self, view, input_node):
+        """Build the validation node for feature validation."""
+        # TODO: Implement validation logic
+        return input_node
+
+    def _build(self, view, input_nodes: Optional[List[DAGNode]]) -> DAGNode:
+        has_physical_source = (hasattr(view, "batch_source") and view.batch_source) or (
+            hasattr(view, "source")
+            and view.source
+            and not isinstance(view.source, FeatureView)
+        )
+
+        is_derived_view = hasattr(view, "source_views") and view.source_views
+
+        if has_physical_source and not is_derived_view:
+            last_node = self.build_source_node(view)
+            if self._should_transform(view):
+                last_node = self.build_transformation_node(view, [last_node])
+        elif input_nodes:
+            if self._should_transform(view):
+                last_node = self.build_transformation_node(view, input_nodes)
+            else:
+                last_node = self.build_join_node(view, input_nodes)
+        else:
+            raise ValueError(f"FeatureView {view.name} has no valid source or inputs")
+
+        if last_node is None:
+            raise ValueError(f"Failed to build processing node for {view.name}")
+
+        last_node = self.build_filter_node(view, last_node)
+
+        if self._should_aggregate(view):
+            last_node = self.build_aggregation_node(view, last_node)
+        elif self._should_dedupe(view):
+            last_node = self.build_dedup_node(view, last_node)
+
+        if self._should_validate(view):
+            last_node = self.build_validation_node(view, last_node)
+
+        return last_node
+
+    def build(self) -> ExecutionPlan:
+        """Build execution plan with support for derived feature views and sink_source writing."""
+        if self.is_historical_retrieval and self._should_aggregate(self.feature_view):
+            return self._build_aggregation_optimized_plan()
+
+        if self.is_materialization:
+            return self._build_materialization_plan()
+
+        return super().build()
+
+    def _build_materialization_plan(self) -> ExecutionPlan:
```

Comment:
Why keep this implementation at the Ray level? Feesl more generic, no?

Reply:
Oh I see. Can we make an Issue to follow up and extend this to make it more generic? As @HaoXuAI mentioned, this approach can be used in Spark as well.